### PR TITLE
Override module chrome with a plugin

### PR DESCRIPTION
**Pull request by jrseliga**
#### Purpose

Template providers often ship the same module chromes with every template. It is much more convenient to manage one file and ship a plugin or a plugin + library that could register these module chromes.

Unfortunately a popular template framework provider already implements this concept by overriding
the core file /libraries/cms/module/helper.php with their framework initialization plugin for the purposes of adding their own triggerEvent.
#### How it works

A plugin that executes on the **onGetModuleChromes** must return an array with keys **element** (friendly name for their list of module chromes) and **path** (path to file containing module chrome definitions)
#### Testing
1. Install Joomla (sample data not necessary)
2. [Upload plugin folders to /plugins/system](https://github.com/jrseliga/module-chrome-plugins) 
3. Discover / Enable plugins **modulechromesone** and **modulechromestwo**
4. Navigate to front-end notice the "Login Form" module
   default-install
5. Module Manager
   - Login Form
     1. Advanced tab
     2. Module Style
        - Scroll to bottom, will see two new grouped lists **Plugin One** and **Plugin Two**
          module-style
   - Choose any from either of these groups
   - Refresh front-end, the name of the module chrome function used will be displayed in place of the log in form.one
   - Repeat last two steps to see different chromes rendered. four

### CODE REVIEW

**Comment by Bakual** [comment]
First you need to fix some codestyle issue. Between the `if` and the opening `(` there has to be a space. So instead of `if(...)` make it `if (...)`.

To get an idea how this should work and to be able to test it, can you add an example plugin which would contain a few overrides?

Also, please change the order. The template override always should have the last word. With your proposal the plugin would override the template override which would be wrong.

**Comment by zero-24** [comment]
@jrseliga @Bakual here is a PR that should fix the described issues and also with some simplications: https://github.com/jrseliga/joomla-cms/pull/1

**Comment by jrseliga** [comment]
@Bakual @zero-24 forgetting $app was a poor copy-paste from my development site.

Looks like @zero-24 fixed the issues, I'll merge that PR.

@Bakual, I will provide a plugin for testing, first I'd like to update /libraries/cms/form/field/chromestyle.php so that the module's configuration can see these chromes as options in "Module Style".

**Comment by jrseliga** [comment]
Originally this was also going to include the ability to override a layout for modules via a plugin, however testing revealed the dependency this has on plugin priority. Injecting a layout in this way is not conducive, or predictable for developers or site administrators. Will remove this event.

Despite the branch name this will only focus on module chromes as the general problem I'd like to solve is template providers needing to change manage the same /templates/template/html/modules.php across all of their templates. If you reuse the same chrome, ship your template with a library and a plugin (or just a plugin) that registers your a path to your chrome file(s). If it matches the Joomla! chrome function naming convention it will be registered.

**Comment by jrseliga** [comment]
Layout override via plugin removed for reason above. Will supply a plugin for testing.

File: libraries/cms/module/helper.php [hunk 1]
```diff
@@ -198,17 +198,23 @@ public static function renderModule($module, $attribs = array())
...
 		}
 
-		include_once JPATH_THEMES . '/system/html/modules.php';
-		$chromePath = JPATH_THEMES . '/' . $template . '/html/modules.php';
+		// Load system chromes and active template chromes
+		$chromePaths = array(JPATH_THEMES . '/system/html/modules.php', JPATH_THEMES . '/' . $template . '/html/modules.php');
 
-		if (!isset($chrome[$chromePath]))
+		// Load chromes from plugins
+		$chromePaths = array_merge($chromePaths, $app->triggerEvent('onGetModuleChromePaths'));
```
Comment (Bakual) [inline libraries/cms/module/helper.php]: I'd first check if the plugin returns indeed a valid array of paths before merging it.
Also you need to think about what happens when there is more than one plugin returning a path.
Maybe you need to pass the plugins an array which will be filled with custom paths.

**Comment by jrseliga** [comment]
@Bakual significant changes were made, the original comment will be rewritten with instructions and more detailed information.

I have two plugins ready for testing to prove it can handle multiple, and have successfully tested functionality.
Should I upload them as a separate repository or is there a way to include them in this branch without the PR noticing them?

**Comment by Bakual** [comment]
I would just put the plugins somewhere available for download. Like Dropbox or your own website. The put the link into the PR here.

File: libraries/cms/module/helper.php [hunk 2]
```diff
@@ -278,35 +302,30 @@ public static function renderModule($module, $attribs = array())
...
-		}
-		elseif (file_exists($bPath))
+		foreach ($layoutPaths as $layoutPath)
 		{
-			return $bPath;
-		}
-		else
-		{
-			return $dPath;
+			if (file_exists($layoutPath))
```
Comment (phproberto) [inline libraries/cms/module/helper.php]: You should use `JPath::find()` here

**Comment by phproberto** [comment]
While I like the idea if at some point `JLayout` is going to be used for this it's probably a good idea to wait for it. This is a new feature so v3.5 is the minimum release for it anyway.

I see a lof of file_exists() checks that could be avoided with `JLayout` or at least with `JPath`

**Comment by nikosdion** [comment]
I love the concept and I agree with Roberto. Using JLayout would be the future-proof way to implement such a fundamental feature, especially considering the key role JLayout is poised to have in Joomla! 4 (at least according to the _informal_ discussion we had with half of the PLT and some Joomla! contributors during the Make It Happen session at JaB 15).

Roberto can you help Justin implement that? You're the de facto JLayout expert, the rest of us are mere apprentices :)

**Comment by jrseliga** [comment]
@phproberto would like to see this available in 3.5 and would be happy to help implement the concept with `JLayout`.

**Comment by hardiktailored** [comment]
I have tested this item :white_check_mark: successfully on d90e033fc3cee41c4929f7b84eaff9f919e66eb2<br /><br />Followed the testing steps and working as explained.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/7099">issues.joomla.org/joomla-cms/7099</a>.</sub>

**Comment by roland-d** [comment]
@jrseliga Could you update this PR to use JLayout? Thank you.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/7099">issues.joomla.org/joomla-cms/7099</a>.</sub>

**Comment by BhaveshTailored** [comment]
I have tested this item :white_check_mark: successfully on d90e033fc3cee41c4929f7b84eaff9f919e66eb2<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/7099">issues.joomla.org/joomla-cms/7099</a>.</sub>

**Comment by brianteeman** [comment]
I am closing this at this time as it appears to have been abandoned by @jrseliga 
It can always be re-opened if updated as per @roland-d comments above

**Pull request closed without merging**